server/src/test/java/org/elasticsearch/rest/RestControllerTests.java

```diff
@@ -623,6 +623,39 @@ public void testDispatchRestrictSystemIndices() {
         assertFalse(context.isSystemIndexAccessAllowed());
     }
 
+    public void testDispatchCompatibleHandler() {
+        final String mimeType = randomFrom("application/vnd.elasticsearch+json;compatible-with=7");
+        String content = randomAlphaOfLength((int) Math.round(BREAKER_LIMIT.getBytes() / inFlightRequestsBreaker.getOverhead()));
+        final List<String> contentTypeHeader = Collections.singletonList(mimeType);
+        FakeRestRequest fakeRestRequest = new FakeRestRequest.Builder(NamedXContentRegistry.EMPTY)
+            .withContent(new BytesArray(content), RestRequest.parseContentType(contentTypeHeader)).withPath("/foo")
+            .withHeaders(Map.of("Content-Type", contentTypeHeader, "Accept", contentTypeHeader))
+            .build();
```

Comment:
Also, what happens if you set the Content-Type to application/json, but Accept to application/vnd.elasticsearch+json;compatible-with=7 ?

I wonder if we need more testing for these edge cases ?(either unit or integration)

Reply:
good question. we can be more specific for response generation with Accept and request parsing with Content-Type.
But on the other hand I think it will just complicate matters too much. 
I would prefer to assume (and assert in code) that these are the same

Reply:
Same as above (I added a "parking lot" section on #51816 so we don't forget.)